### PR TITLE
re-enable dead letter logging after specified duration, #24874

### PR DESCRIPTION
**Pull request by patriknw**
Completing PR started in https://github.com/akka/akka/pull/25131

Refs #24874

### CODE REVIEW

File: akka-actor/src/main/scala/akka/event/DeadLetterListener.scala [hunk 1]
```diff
@@ -26,21 +32,72 @@ class DeadLetterListener extends Actor {
...
+      incrementCount()
+      if (count == maxCount) {
+        logDeadLetter(message, snd, recipient, ", no more dead letters will be logged")
+        context.stop(self)
+      } else {
+        logDeadLetter(message, snd, recipient, "")
+      }
+  }
+
+  private def receiveWithSupsendLogging(suspendDuration: FiniteDuration): Receive = {
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/event/DeadLetterListener.scala]:
```suggestion
  private def receiveWithSuspendLogging(suspendDuration: FiniteDuration): Receive = {
```

File: akka-actor/src/main/scala/akka/event/DeadLetterListener.scala [hunk 2]
```diff
@@ -26,21 +32,72 @@ class DeadLetterListener extends Actor {
...
-          "This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' " +
-          "and 'akka.log-dead-letters-during-shutdown'."))
-      if (done) context.stop(self)
+  }
+
+  def receive: Receive =
+    if (isAlwaysLoggingDeadLetters) receiveWithAlwaysLogging
+    else
+      context.system.settings.LogDeadLettersSuspendDuration match {
+        case suspendDuration: FiniteDuration => receiveWithSupsendLogging(suspendDuration)
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/event/DeadLetterListener.scala]:
```suggestion
        case suspendDuration: FiniteDuration => receiveWithSuspendLogging(suspendDuration)
```

File: akka-actor/src/main/scala/akka/event/DeadLetterListener.scala [hunk 3]
```diff
@@ -26,21 +32,72 @@ class DeadLetterListener extends Actor {
   override def postStop(): Unit =
     eventStream.unsubscribe(self)
 
-  def receive = {
-    case DeadLetter(message, snd, rcp) =>
+  private def incrementCount(): Unit = {
+    // `count` is public API (for unknown reason) so for backwards compatibility reasons we
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/event/DeadLetterListener.scala]: So, we expect that people extend and adapt this class?

File: akka-actor/src/main/scala/akka/event/DeadLetterListener.scala [hunk 4]
```diff
@@ -26,21 +32,72 @@ class DeadLetterListener extends Actor {
...
+  }
+
+  private def receiveWhenSuspended(suspendDuration: FiniteDuration, suspendDeadline: Deadline): Receive = {
+    case DeadLetter(message, snd, recipient) =>
+      incrementCount()
+      if (suspendDeadline.isOverdue()) {
+        val doneMsg = s", of which ${count - maxCount - 1} were not logged. The counter will be reset now"
+        logDeadLetter(message, snd, recipient, doneMsg)
+        count = 0
+        context.become(receiveWithSupsendLogging(suspendDuration))
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/event/DeadLetterListener.scala]:
```suggestion
        context.become(receiveWithSuspendLogging(suspendDuration))
```

File: akka-actor/src/main/scala/akka/event/DeadLetterListener.scala [hunk 5]
```diff
@@ -26,21 +32,72 @@ class DeadLetterListener extends Actor {
   override def postStop(): Unit =
     eventStream.unsubscribe(self)
 
-  def receive = {
-    case DeadLetter(message, snd, rcp) =>
+  private def incrementCount(): Unit = {
+    // `count` is public API (for unknown reason) so for backwards compatibility reasons we
+    // can't change it to Long
+    if (count == Int.MaxValue)
+      count = 1
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/event/DeadLetterListener.scala]: Maybe log here?

**Review by jrudolph — COMMENTED** [review]
A few suggestions below, otherwise, LGTM

**Comment by patriknw** [comment]
incorporated in 5f46fdba41

File: akka-actor/src/main/scala/akka/event/DeadLetterListener.scala (hunk 3 above)
Comment (patriknw) [reply akka-actor/src/main/scala/akka/event/DeadLetterListener.scala]: it shouldn't have been public in the first place, but now it is

**Review by chbatey — APPROVED** [review]
Nice, LGTM

**Comment by patriknw** [comment]
Thanks @hepin1989 for the original initiative

**Pull request merged**